### PR TITLE
v1.3.4

### PR DESCRIPTION
**Pull request by StanfordChiang**
bug fix - once a token was cached, the sebsequent tokens from the same client were not validated properly

### CODE REVIEW

File: lib/key.js [hunk 1]
```diff
@@ -376,21 +376,24 @@ module.exports = function (keyOptions, clientId, secret, self) {
...
-			if (checkToken && checkToken.expires >= new Date()) {
-				return resolve(true);
-			} else {
-				// ensure the tokenMap is cleared for the specified clientId
-				tokenMap.delete(clientId);
+			// determine if a valid token exists if cacheTokens is enabled
+			if (settings.cacheTokens) {
+				let cachedToken = tokenMap.get(clientId);
+				if (cachedToken) {
+					if (cachedToken.tokenId === (token === 'string' ? token : token.tokenId)) {
```
Comment (tsmith1985) [inline lib/key.js]: @StanfordChiang should be `typeof token === 'string'` right?

File: test/lib/key.js [hunk 2]
```diff
@@ -913,6 +913,38 @@ describe('key', () => {
...
+					// tokenCacheSize should be still 1 because we use the same client
+					key.getTokenCacheSize().should.equal(1);
+
+					// badtoken should not be valid
+					valid2.should.be.false;
+
+					return done();
+				});
+			});
+		});
```
Comment (tsmith1985) [inline test/lib/key.js:947]: @StanfordChiang I would recommend adding a test that validates that a valid cached token can be used i.e. ensure that a subsequent request to the key-api is avoided if a valid cached token exists.

**Review by tsmith1985 — COMMENTED** [review]
@StanfordChiang think I found one issue. see comments.

File: lib/key.js (hunk 1 above)
Comment (StanfordChiang) [reply lib/key.js]: i'll fix that.  thanks @tsmith1985

File: history.md [hunk 3]
```diff
@@ -1,3 +1,11 @@
+# v1.3.4
+
+* Bug Fix - addressed an issue once a client's valid authentication token was cached, subsequent tokens from the same client were not validate properly
+
+# v1.3.3
```
Comment (brozeph) [inline history.md:5]: It looks like `1.3.1` and `1.3.2` were to add support for provisioning routes and `1.3.3` was to handle expired token refresh on reconnect - would it be possible to add notes to the history for this?

File: lib/key.js [hunk 4]
```diff
@@ -376,21 +376,24 @@ module.exports = function (keyOptions, clientId, secret, self) {
...
-				};
-
-			// check to see if the token has already been validated...
-			if (checkToken && checkToken.expires >= new Date()) {
-				return resolve(true);
-			} else {
-				// ensure the tokenMap is cleared for the specified clientId
-				tokenMap.delete(clientId);
+			// determine if a valid token exists if cacheTokens is enabled
+			if (settings.cacheTokens) {
```
Comment (brozeph) [inline lib/key.js:380]: A style thing to remove nested if statements, but definitely a low priority adjustment:

```javascript
if (settings.cacheTokens && tokenMap.has(clientId)) {
	let cachedToken = tokenMap.get(clientId);
	if (cachedToken.tokenId === (typeof token === 'string' ? token : token.tokenId)) {
		if (cachedToken.expires >= new Date()) {
			return resolve(true);
		}

		tokenMap.delete(clientId);
	}
}
```

**Comment by StanfordChiang** [comment]
@brozeph @tsmith1985 Thank you for reviewing and feedback; I've made necessary change/enhancement.  Please let a look and let me know with any other thoughts/comments.  Thank you

File: lib/key.js [hunk 5]
```diff
@@ -376,21 +376,24 @@ module.exports = function (keyOptions, clientId, secret, self) {
...
+				let cachedToken = tokenMap.get(clientId);
+				if (cachedToken) {
+					if (cachedToken.tokenId === (typeof token === 'string' ? token : token.tokenId)) {
+						if (cachedToken.expires >= new Date()) {
+							return resolve(true);
+						}
+						// ensure the tokenMap is cleared when the token has expired
+						tokenMap.delete(clientId);
+					}
+				}
```
Comment (tsmith1985) [inline lib/key.js:390]: @StanfordChiang Looks good!

File: test/lib/key.js (hunk 2 above)
Comment (tsmith1985) [reply test/lib/key.js:947]: @StanfordChiang Nice!

**Review by tsmith1985 — APPROVED** [review]
@StanfordChiang Looks good!

**Pull request merged**